go/extractor/project/project.go

```diff
@@ -42,109 +124,366 @@ func checkDirsNested(inputDirs []string) (string, bool) {
 	return dirs[0], true
 }
 
-// Returns the directory to run the go build in and whether to use a go.mod
-// file.
-func findGoModFiles(emitDiagnostics bool) (baseDir string, useGoMod bool) {
-	goModPaths := util.FindAllFilesWithName(".", "go.mod", "vendor")
-	if len(goModPaths) == 0 {
-		baseDir = "."
-		useGoMod = false
+// A list of files we created that should be removed after we are done.
+var filesToRemove []string = []string{}
+
+// Try to initialize a go.mod file for projects that do not already have one.
+func InitGoModForLegacyProject(path string) {
+	log.Printf("Project appears to be a legacy Go project, attempting to initialize go.mod in %s\n", path)
+
+	modInit := toolchain.InitModule(path)
+
+	if !util.RunCmd(modInit) {
+		log.Printf("Failed to initialize go.mod file for this project.")
 		return
 	}
-	goModDirs := getDirs(goModPaths)
-	if util.AnyGoFilesOutsideDirs(".", goModDirs...) {
-		if emitDiagnostics {
-			diagnostics.EmitGoFilesOutsideGoModules(goModPaths)
+
+	// Add the go.mod file to a list of files we should remove later.
+	filesToRemove = append(filesToRemove, filepath.Join(path, "go.mod"))
+
+	modTidy := toolchain.TidyModule(path)
+	out, err := modTidy.CombinedOutput()
+	log.Println(string(out))
+
+	if err != nil {
+		log.Printf("Failed to determine module requirements for this project.")
+	}
+
+	if strings.Contains(string(out), "is relative, but relative import paths are not supported in module mode") {
+		diagnostics.EmitRelativeImportPaths()
```

Comment:
Interesting! I thought relative import paths did work in module mode. As that seems to not be the case, we should update the message for this diagnostic and remove the suggestion to use a Go module.